### ISSUE TITLE
[meta] use tools like `changesets` for better maintainance

### ISSUE THREAD

**Issue by JounQin** [Sec-ant/prettier-plugin-embed#55]
https://github.com/changesets/changesets

**Comment by Sec-ant** [comment]
Closed by #75

**Comment by JounQin** [comment]
It seems `changeset publish` is not used in CI for automatic release.

**Comment by Sec-ant** [comment]
> It seems `changeset publish` is not used in CI for automatic release.

Done

**Comment by JounQin** [comment]
I still don't understand why `npm pack` or publish artifact are required, `changesets` will handle `npm publish` and `git push --follow-tags` and GitHub releases already.

**Comment by Sec-ant** [comment]
> I still don't understand why `npm pack` or publish artifact are required, `changesets` will handle `npm publish` and `git push --follow-tags` and GitHub releases already.
> 
> What means you don't need a custom `version` input at all.

That's just for debugging or tests, they're only triggered by workflow dispatch events (manually). A pull event on main branch will still let changeset automatically handle everything.

**Comment by JounQin** [comment]
The best practice for changesets is merging final MR.

And they still in workflow definition file, what does debug and test mean here then?

**Comment by Sec-ant** [comment]
> The best practice for changesets is merging final MR.

Sure, but this workflow doesn't change the changesets behavior.

> And they still in workflow definition file.

These steps will be skipped if the workflow is not triggered by workflow dispatch.

> what does debug and test mean here then?

Like I can manually trigger a package building process and download it from the artifacts to see everything works as expected? It's just a little manual control over the repo, I don't expect I will use that a lot.

**Comment by JounQin** [comment]
Isn't that should be covered by Codesandbox CI or https://github.com/un-ts/preversion?

**Comment by Sec-ant** [comment]
> Isn't that should be covered by Codesandbox CI or https://github.com/un-ts/preversion?

CodeSandBox CI works great on testing package before the PR is to be merged, so I'm definitely not going to reinvent the wheels.

The difference is that I prefer to have a manually triggered mechanism to check everything works without doing some intrusive actions (without actually publishing it), at an arbitrary time point (doesn't have to be triggered by PR or pushes), with just one click. Sometimes, I also want to make sure the package can be successfully built on an average machine.

And despite the easiness of CodeSandBox-CI, there're scenarios in which it will have problems: In a repo which uses Emscripten to compile the Cpp source code into WASM files before doing a full build; In a repo which requires Biome or other non-js tools to process the source code files.

**Comment by JounQin** [comment]
I don't understand why a manually triggered workflow works differently with Codesandbox CI or https://github.com/un-ts/preversion/actions/workflows/release-beta.yml, you can also manually trigger the workflow.

<img width="1378" alt="image" src="https://github.com/Sec-ant/prettier-plugin-embed/assets/8336744/fd267823-58e8-4d15-8f09-192bc3c36332">

**Comment by Sec-ant** [comment]
> I don't understand why a manually triggered workflow works differently with Codesandbox CI.

Because I cannot manually trigger a CodeSandBox CI unless I start a pull request? which is too much an operation and I consider it intrusive. And regarding the reasons I listed in https://github.com/Sec-ant/prettier-plugin-embed/issues/55#issuecomment-1913881207, there're some tools that cannot be installed in CodeSandBox CI.

> https://github.com/un-ts/preversion/actions/workflows/release-beta.yml, you can also manually trigger the workflow.

Exactly, so I just wrote my own `release-beta.yml` without actually publishing it to NPM. As I don't expect to use this frequently and the steps are trivial, I don't see the necessity to introduce another tool or spend time to find one.

This is just for myself to play around with the package, like to test it in another project during dev. I'd rather not put much effort to standardize or generalize the workflow.
